### PR TITLE
Added some cases of guest agent

### PR DESCRIPTION
**Pull request by iccaszhulili**
1) Added some cases of guest agent about guest agent state
check after restart libvirtd or guest life-cycle operation.
2) Added some cases of guest agent about hotunplug guest agent

Signed-off-by: Lili Zhu <lizhu@redhat.com>

### CODE REVIEW

**Comment by iccaszhulili** [comment]
# avocado run --vt-type libvirt guest_agent
JOB ID     : 7abc8d71d370e4f9983eef41b9b58fb734cc86ad
JOB LOG    : /root/avocado/job-results/job-2020-12-18T03.27-7abc8d7/job.log
 (1/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.ga_state_test: PASS (50.66 s)
 (2/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.restart_libvirtd: PASS (51.38 s)
 (3/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.suspend_resume_guest: PASS (53.28 s)
 (4/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.hotunplug_ga: PASS (50.55 s)
 (5/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.customize_source_path: PASS (51.88 s)
 (6/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.ga_state_test: PASS (109.45 s)
 (7/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.restart_libvirtd: PASS (51.31 s)
 (8/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.suspend_resume_guest: PASS (50.84 s)
 (9/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.customize_source_path: PASS (51.78 s)
RESULTS    : PASS 9 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 523.73 s

File: libvirt/tests/src/guest_agent.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        cur_xml = vm_xml.VMXML.new_from_dumpxml(vm_name)
+        channels = cur_xml.get_devices('channel')
+        for channel in channels:
+            target = channel['xmltreefile'].find('./target')
+            if target is not None:
+                name = target.get('name')
+                if name and name.startswith("org.qemu.guest_agent"):
+                    ga_xml = channel.xml
+                    break
+        return ga_xml
```
Comment (chloerh) [inline libvirt/tests/src/guest_agent.py:48]: `ga_xml` could be undefined. How about we init it first?

File: libvirt/tests/src/guest_agent.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        if prepare_channel:
+            vm.prepare_guest_agent(start=start_ga, channel=True,
+                                   source_path=src_path)
+
+        if restart_libvirtd:
+            utils_libvirtd.libvirtd_restart()
+
+        if suspend_resume_guest:
+            virsh.suspend(vm_name, ignore_status=True, debug=True)
+            virsh.resume(vm_name, ignore_status=True, debug=True)
```
Comment (chloerh) [inline libvirt/tests/src/guest_agent.py:85]: Why we ignore status? We don't care whether commands fail?

File: libvirt/tests/src/guest_agent.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+from virttest import utils_selinux
+from virttest.utils_test import libvirt
+from virttest.libvirt_xml import vm_xml
+
+
+def run(test, params, env):
+    def check_ga_state(start_ga):
+        # The session is just to make sure the guest
+        # is fully boot up
+        session = vm.wait_for_login()
```
[the remark this answers is not in the document]
Comment (iccaszhulili) [reply libvirt/tests/src/guest_agent.py]: updated

File: libvirt/tests/src/guest_agent.py (hunk 1 above)
Comment (iccaszhulili) [reply libvirt/tests/src/guest_agent.py:48]: updated

File: libvirt/tests/src/guest_agent.py (hunk 2 above)
Comment (iccaszhulili) [reply libvirt/tests/src/guest_agent.py]: @chloerh updated. Thanks

File: libvirt/tests/src/guest_agent.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+from virttest import utils_selinux
+from virttest.utils_test import libvirt
+from virttest.libvirt_xml import vm_xml
+
+
+def run(test, params, env):
+    def check_ga_state(start_ga):
+        # The session is just to make sure the guest
+        # is fully boot up
+        session = vm.wait_for_login().close()
```
Comment (chloerh) [inline libvirt/tests/src/guest_agent.py:16]: I guess you forgot to remove `session=`

**Review by chloerh — COMMENTED** [review]
Only 1 comment, the rest is ok to me.

File: libvirt/tests/src/guest_agent.py (hunk 4 above)
Comment (kamilvarga) [reply libvirt/tests/src/guest_agent.py:16]: Agree with @chloerh.

**Review by kamilvarga — COMMENTED** [review]
As discussed in other review, would be better to put functions outside the run(). Other than that and comments from Chloerh it looks good to me.

**Comment by iccaszhulili** [comment]
@chloerh @chloerh Updated 
And please do not merge it before this PR
https://github.com/avocado-framework/avocado-vt/pull/2862

Thanks

**Comment by chloerh** [comment]
According to the result of travis check, seems the code could use some modifications....I'll just remove the review request. But please, feel free to request again.

**Comment by iccaszhulili** [comment]
@chloerh updated, and added the latest testing result

JOB ID     : 95a693044001cf1b613a22b9efc50c1ee1b6a705
JOB LOG    : /root/avocado/job-results/job-2021-04-12T00.57-95a6930/job.log
 (1/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.ga_state_test: PASS (54.34 s)
 (2/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.restart_libvirtd: PASS (55.44 s)
 (3/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.suspend_resume_guest: PASS (56.02 s)
 (4/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.hotunplug_ga: PASS (71.34 s)
 (5/9) type_specific.io-github-autotest-libvirt.guest_agent.positive.ga_started.customize_source_path: PASS (56.72 s)
 (6/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.ga_state_test: PASS (54.51 s)
 (7/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.restart_libvirtd: PASS (55.19 s)
 (8/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.suspend_resume_guest: PASS (56.23 s)
 (9/9) type_specific.io-github-autotest-libvirt.guest_agent.negative.ga_stopped.customize_source_path: PASS (55.24 s)
RESULTS    : PASS 9 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 516.76 s

**Review by chloerh — COMMENTED** [review]
The change is ok to me. But we'd better provide docstring for our functions. 
Please refer to functions in https://github.com/autotest/tp-libvirt/pull/3384/files

**Comment by iccaszhulili** [comment]
@chloerh updated, thanks very much

**Review by chloerh — APPROVED** [review]
All checked. Very nice.

**Pull request merged**